### PR TITLE
Expose kvs and ts and tls when not initialized

### PR DESCRIPTION
**Pull request by fsamuel-bs**
**Goals (and why)**: Other products might need them for their initialization steps, so we shouldn't throw when accessing them.

**Implementation Description (bullets)**: -

**Concerns (what feedback would you like?)**: -

**Where should we start reviewing?**: -

**Priority (whenever / two weeks / yesterday)**: -

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/2564)

### CODE REVIEW

File: docs/source/release_notes/release-notes.rst [hunk 1]
```diff
@@ -135,6 +135,11 @@ develop
           If you used one of the deleted or deprecated constructors, use the static ``create`` method.
           (`Pull Request <https://github.com/palantir/atlasdb/pull/2549>`__)
 
+    *   - |improved|
+        - Expose the TimestampService, KeyValueService and TimeLockService before initialization when initializing asynchronously.
+          AtlasDB users might need these objects when initializing, so we need to expose them without throwing when not initialized.
+          (`Pull Request <https://github.com/palantir/atlasdb/pull/2564>`__)
```
Comment (jeremyk-91) [inline docs/source/release_notes/release-notes.rst]: Probably common-sense - but the methods of these objects will throw a `NotInitializedException` if users try to invoke them, right? Might be worth mentioning

**Comment by jeremyk-91** [comment]
Actually, is there a risk we might run into some kind of unsafe-publication scenario, where people are able to use the KVS before the tables have been created successfully?

**Comment by fsamuel-bs** [comment]
We're exposing the wrapped objects here. If someone invokes a method on them, it'll just throw.

**Comment by jeremyk-91** [comment]
Ah okay. What I meant was that I was concerned about a state where say the KVS itself has been initialized (e.g. Cassandra is up and we've created the keyspace), but some of the later steps (e.g. creating the tables from the schema) haven't taken place yet. Whereas if you wait for the TM to be initialized these are going to be there. @nziebart any thoughts here?

**Comment by fsamuel-bs** [comment]
In general, I agree that we shouldn't be exposing the KVS. The motivation for this PR is that an internal storage product calls `kvs.getAllTableNames()` on its health-check. This can (and should) be replaced by an Atlas health-check, when we implement it. Therefore we might be fine with not exposing the KVS in this scenario.

When talking to @jeremyk-91 this morning, we notice that we might still have problems if we expose the timestamp service. It might be the case that we did not migrate the timestamp bound, but the timestamp service declares itself as initialized. We don't want to have calls to the TimestampService succeeding in this case, as we would effectively go back in time before the migration actually succeeds.

**Comment by nziebart** [comment]
Yeah the problem here is that, these objects are not actually ready for use until some external logic has been run on them. On a related note, are we confident that internal usages of these objects (potentially pre-initialization) are safe?

**Comment by hsaraogi** [comment]
This blocks releasing async initialize as a usable feature.

**Comment by gsheasby** [comment]
Are we sure of that? I think we should probably try to use async initialisation on the internal product that we manage, and verify what we actually need to expose (it could be a proper subset of these three things)

**Comment by tboam** [comment]
@fsamuel-bs to resolve conversation here and make any necessary changes.

**Comment by fsamuel-bs** [comment]
After reviewing the code from the internal storage product we maintain, I'm convinced that this is not needed anymore. Closing for now.

**Pull request closed without merging**